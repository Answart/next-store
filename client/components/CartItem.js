import React from 'react';
import Link from 'next/link';
import PropTypes from 'prop-types';
import PriceTag from './PriceTag';
import SvgIcon from './SvgIcon';
import { RemoveFromCart, UpdateCartItem } from './Buttons';
import { formatMoney, capWord } from '../lib/utils';


const CartItem = ({ id, quantity, variant }) => {
  if (!variant)
    return (
      <tr>
        <td className="cart-item-details">
          <p>This Product has been removed</p>
        </td>

        <td className="cart-item-quantity">
          <div className="cart-item-quantity-remove">
            <RemoveFromCart id={id} />
          </div>
        </td>
      </tr>
    );
  const title = variant.product.title;
  const price = variant.sale
    ? variant.salePrice
    : variant.price;
  return (
    <tr>
      <td className="cart-item-details">
        <div className='cart-item-image'>
          <Link href={{
            pathname: '/shop',
            query: { id: variant.product.id }
          }}>
            <a>
              <img width="100" height="120"
                src={variant.image.image_url}
                alt={title}
                title={`Go to: ${title}`}
              />
            </a>
          </Link>
        </div>

        <div className='cart-item-meta'>
          <div className="cart-item-detail">
            <h4>{title}</h4>
          </div>

          <div className="cart-item-detail">
            <PriceTag
              price={variant.price}
              sale={variant.sale}
              salePrice={variant.salePrice}
            />
          </div>

          {variant.size && (
            <div className="cart-item-detail">
              Size: {variant.size}
            </div>
          )}

          {variant.color && (
            <div className="cart-item-detail">
              Color: {capWord(variant.color)}
            </div>
          )}
        </div>
      </td>

      <td className="cart-item-quantity">
        <div className="cart-item-quantity-actions">
          <UpdateCartItem
            id={variant.id}
            quantity={quantity - 1}
            disabled={(quantity - 1) <= 0}
          >
            <SvgIcon width={7} name='minus' color='darkBlue' />
          </UpdateCartItem>

          {quantity}

          <UpdateCartItem
            id={variant.id}
            quantity={quantity + 1}
            disabled={(quantity + 1) > variant.quantity}
          >
            <SvgIcon width={7} name='plus' color='darkBlue' />
          </UpdateCartItem>

          <div className="cart-item-quantity-remove">
            <RemoveFromCart id={id} />
          </div>
        </div>
      </td>

      <td className="cart-item-total-price">
        {formatMoney(price * quantity)}
      </td>
    </tr>
  );
};

CartItem.propTypes = {
  id: PropTypes.string.isRequired,
  quantity: PropTypes.number.isRequired,
  variant: PropTypes.shape({
    price: PropTypes.number.isRequired,
    quantity: PropTypes.number.isRequired,
    size: PropTypes.string.isRequired,
    color: PropTypes.string.isRequired,
    sale: PropTypes.bool.isRequired,
    salePrice: PropTypes.number.isRequired,
    image: PropTypes.shape({
      id: PropTypes.string.isRequired,
      image_url: PropTypes.string.isRequired
    }),
    product: PropTypes.shape({
      id: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired
    }).isRequired
  })
};


export default CartItem;
