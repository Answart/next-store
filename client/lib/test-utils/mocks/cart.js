import {
  ADD_TO_CART_MUTATION,
  UPDATE_CARTITEM_MUTATION,
  REMOVE_FROM_CART_MUTATION
} from '../../../graphql'
import { fakeCartItem } from '../utils';
import { mockVariant } from './variant';

const mockCartItem = fakeCartItem();


const addToCartMutationMock = {
  request: { query: ADD_TO_CART_MUTATION, variables: { id: mockVariant.id } },
  result: {
    data: {
      addToCart: {
        __typename: mockCartItem.__typename,
        id: mockCartItem.id,
      },
    },
  },
};

const updateCartItemMutationMock = (quantity) => ({
  request: { query: UPDATE_CARTITEM_MUTATION, variables: { id: mockCartItem.id, quantity } },
  result: {
    data: {
      updateCartItem: {
        __typename: mockCartItem.__typename,
        id: mockCartItem.id,
        quantity,
        variant: { ...mockCartItem.variant }
      },
    },
  },
});

const removeFromCartMutationMock = {
  request: { query: REMOVE_FROM_CART_MUTATION, variables: { id: mockCartItem.id } },
  result: {
    data: {
      removeFromCart: {
        __typename: mockCartItem.__typename,
        id: mockCartItem.id,
      },
    },
  },
};


export {
  mockCartItem,
  addToCartMutationMock,
  updateCartItemMutationMock,
  removeFromCartMutationMock
}
