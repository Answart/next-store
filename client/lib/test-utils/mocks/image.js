import { CREATE_IMAGE_MUTATION } from '../../../graphql'
import { fakeImage } from '../utils';

const mockImage = fakeImage();
const mockImageVariables = {
  cloudinary_id: mockImage.cloudinary_id,
  name: mockImage.name,
  height: mockImage.height,
  width: mockImage.width,
  transformation: mockImage.transformation,
  image_url: mockImage.image_url,
  large_image_url: mockImage.large_image_url
}


const createImageMutationMock = (overrides) => ({
  request: {
    query: CREATE_IMAGE_MUTATION,
    variables: {
      ...mockImageVariables,
      ...overrides,
    },
  },
  result: {
    data: {
      createImage: {
        __typename: 'Image',
        id: mockImage.id,
      },
    },
  },
});

const createImageMutationErrorMock = {
  request: {
    query: CREATE_IMAGE_MUTATION,
    variables: {
      ...mockImageVariables,
      cloudinary_id: '',
      name: '',
    },
  },
  result: {
    errors: [{ message: 'ack!' }],
  }
};


export {
  mockImage, mockImageVariables,
  createImageMutationMock, createImageMutationErrorMock,
}
