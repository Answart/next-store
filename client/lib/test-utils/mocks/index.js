export {
  mockUser, mockUsers,
  mockImage, mockImageVariables,
  mockProduct, mockProducts, mockShopProductsVariables,
  mockVariant,
  mockCartItem,
  mockOrderItem, mockOrder,
} from './typeDefs';
export {
  localUserQueryEmptyCartMock, localUserQueryNoUserMock, localUserQueryCartItemMock,
  userQueryEmptyCartMock, userQueryNoUserMock, userQueryCartItemMock,
  signupMutationMock,
  signinMutationMock,
  signoutMutationMock,
  requestPasswordResetMutationMock, requestPasswordResetMutationErrorMock,
  resetPasswordMutationMock, resetPasswordMutationErrorMock,
  updatePermissionsMutationMock,
  createImageMutationMock, createImageMutationErrorMock,
  productQueryMock, productQueryErrorMock, productQueryNoProductMock, productQueryNoVariantMock,
  shopProductsQueryFilterMock,
  createProductMutationMock, createProductMutationErrorMock,
  updateProductMutationMock, updateProductMutationErrorMock,
  deleteProductMutationMock,
  createProductVariantMutationMock, createProductVariantMutationErrorMock,
  updateProductVariantMutationMock, updateProductVariantMutationErrorMock,
  deleteProductVariantMutationMock,
  addToCartMutationMock,
  updateCartItemMutationMock,
  removeFromCartMutationMock,
} from './resolvers';
