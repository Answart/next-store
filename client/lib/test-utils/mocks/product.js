import {
  PRODUCT_QUERY, SHOP_PRODUCTS_QUERY,
  CREATE_PRODUCT_MUTATION,
  UPDATE_PRODUCT_MUTATION,
  DELETE_PRODUCT_MUTATION,
} from '../../../graphql'
import { mockUser, mockProduct, mockImage, mockVariant, mockShopProductsVariables } from '../utils';


const productQueryMock = {
  request: { query: PRODUCT_QUERY, variables: { id: mockProduct.id } },
  result: {
    data: {
      product: {
        ...mockProduct,
        __typename: mockProduct.__typename,
        image: mockImage,
        user: {
          __typename: mockUser.__typename,
          id: mockUser.id,
          name: mockUser.name,
        },
        variants: [{
          ...mockVariant,
          product: {
            __typename: mockProduct.__typename,
            id: mockProduct.id,
            image: mockImage
          }
        }]
      },
    },
  },
};

const productQueryNoVariantMock = {
  request: { query: PRODUCT_QUERY, variables: { id: mockProduct.id } },
  result: {
    data: {
      product: {
        ...mockProduct,
        __typename: mockProduct.__typename,
        image: mockImage,
        user: {
          __typename: mockUser.__typename,
          id: mockUser.id,
          name: mockUser.name,
        },
        variants: []
      },
    },
  },
};

const createProductMutationMock = {
  request: {
    query: CREATE_PRODUCT_MUTATION,
    variables: {
      department: mockProduct.department,
      title: mockProduct.title,
      description: mockProduct.description,
      category: mockProduct.category,
      brand: mockProduct.brand,
      online: mockProduct.online,
      imageId: mockImage.id,
    },
  },
  result: {
    data: {
      createProduct: {
        __typename: 'Product',
        id: mockProduct.id,
      },
    },
  },
};

const createProductMutationErrorMock = {
  request: {
    query: CREATE_PRODUCT_MUTATION,
    variables: {
      department: 'tops',
      title: '',
      description: '',
      category: '',
      brand: '',
      online: false,
      imageId: mockImage.id,
    },
  },
  result: {
    errors: [{ message: 'ack!' }],
  }
};

const updateProductMutationMock = {
  request: {
    query: UPDATE_PRODUCT_MUTATION,
    variables: {
      id: mockProduct.id,
      department: mockProduct.department,
      title: mockProduct.title,
      description: mockProduct.description,
      category: mockProduct.category,
      brand: mockProduct.brand,
      online: mockProduct.online,
      imageId: mockImage.id,
    },
  },
  result: {
    data: {
      updateProduct: {
        __typename: 'Product',
        id: mockProduct.id,
      },
    },
  },
};

const updateProductMutationErrorMock = {
  request: {
    query: UPDATE_PRODUCT_MUTATION,
    variables: {
      id: mockProduct.id,
      department: 'tops',
      title: '',
      description: '',
      category: '',
      brand: '',
      online: false,
      imageId: mockImage.id,
    },
  },
  result: {
    errors: [{ message: 'ack!' }],
  }
};

const deleteProductMutationMock = {
  request: { query: DELETE_PRODUCT_MUTATION, variables: { id: mockProduct.id } },
  result: {
    data: {
      deleteProduct: {
        __typename: 'Product',
        id: mockProduct.id,
      },
    },
  },
};

const shopProductsQueryNameEmptyMock = {
  request: { query: SHOP_PRODUCTS_QUERY, variables: { name: mockUser.name }},
  result: {
    data: {
      products: [],
    },
  },
};

const shopProductsQueryProductMock = {
  request: { query: SHOP_PRODUCTS_QUERY, variables: { ...mockShopProductsVariables }},
  result: {
    data: {
      products: [{
        ...mockProduct,
        image: mockImage,
        user: {
          __typename: mockUser.__typename,
          id: mockUser.id,
          name: mockUser.name,
        },
        variants: [{
          ...mockVariant,
          product: {
            __typename: mockProduct.__typename,
            id: mockProduct.id,
            image: mockImage,
          }
        }]
      }],
    },
  },
};


export {
  productQueryMock, productQueryNoVariantMock,
  shopProductsQueryNameEmptyMock, shopProductsQueryProductMock,
  createProductMutationMock, createProductMutationErrorMock,
  updateProductMutationMock, updateProductMutationErrorMock,
  deleteProductMutationMock,
}
