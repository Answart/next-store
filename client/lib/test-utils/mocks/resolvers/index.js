export {
  localUserQueryEmptyCartMock, localUserQueryNoUserMock, localUserQueryCartItemMock,
} from './local';
export {
  userQueryEmptyCartMock, userQueryNoUserMock, userQueryCartItemMock,
  signupMutationMock,
  signinMutationMock,
  signoutMutationMock,
  requestPasswordResetMutationMock, requestPasswordResetMutationErrorMock,
  resetPasswordMutationMock, resetPasswordMutationErrorMock,
  updatePermissionsMutationMock,
} from './user';
export {
  createImageMutationMock, createImageMutationErrorMock,
} from './image';
export {
  productQueryMock, productQueryErrorMock, productQueryNoProductMock, productQueryNoVariantMock,
  shopProductsQueryFilterMock,
  createProductMutationMock, createProductMutationErrorMock,
  updateProductMutationMock, updateProductMutationErrorMock,
  deleteProductMutationMock,
} from './product';
export {
  createProductVariantMutationMock, createProductVariantMutationErrorMock,
  updateProductVariantMutationMock, updateProductVariantMutationErrorMock,
  deleteProductVariantMutationMock
} from './variant';
export {
  addToCartMutationMock,
  updateCartItemMutationMock,
  removeFromCartMutationMock
} from './cartItem';
