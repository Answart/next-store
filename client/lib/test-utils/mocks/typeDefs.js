import {
  fakeUser, fakeImage, fakeProduct, fakeVariant, fakeCartItem, fakeOrder, fakeOrderItem,
} from '../utils';


const mockUser = fakeUser();
const mockUser2 = fakeUser();
const mockImage = fakeImage();
const mockProduct = fakeProduct();
const mockVariant = fakeVariant();
const mockCartItem = fakeCartItem();
const mockOrder = fakeOrder();
const mockOrderItem = fakeOrderItem();

mockOrder.items.push(mockOrderItem);

const mockUsers = [
  mockUser,
  {
    ...mockUser2,
    id: '87654321',
    permissions: ['USER'],
  }
];
const mockProducts = [{
  ...mockProduct,
  image: mockImage,
  user: {
    __typename: mockUser.__typename,
    id: mockUser.id,
    name: mockUser.name,
  },
  variants: [{
    ...mockVariant,
    product: {
      __typename: mockProduct.__typename,
      id: mockProduct.id,
      image: mockImage,
    }
  }]
}];

const mockImageVariables = {
  cloudinary_id: mockImage.cloudinary_id,
  name: mockImage.name,
  height: mockImage.height,
  width: mockImage.width,
  transformation: mockImage.transformation,
  image_url: mockImage.image_url,
  large_image_url: mockImage.large_image_url
};

const mockShopProductsVariables = {
  name: mockUser.name,
  orderBy: 'createdAt_DESC',
  skip: 0,
  first: 1
 };


export {
  mockUser, mockUsers,
  mockImage, mockImageVariables,
  mockProduct, mockProducts, mockShopProductsVariables,
  mockVariant,
  mockCartItem,
  mockOrder, mockOrderItem,
};
