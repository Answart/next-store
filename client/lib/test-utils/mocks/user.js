import {
  CURRENT_USER_QUERY,
  SIGNOUT_MUTATION,
  REQUEST_PASSWORD_RESET_MUTATION,
} from '../../../graphql'
import { fakeUser } from '../utils';
import { mockCartItem } from './cart';

const mockUser = fakeUser();


const userQueryEmptyCartMock = {
  request: { query: CURRENT_USER_QUERY },
  result: {
    data: {
      me: {
        ...mockUser,
        cart: [],
      },
    },
  },
};

const userQueryNoUserMock = {
  request: { query: CURRENT_USER_QUERY },
  result: {
    data: { me: null }
  },
};

const userQueryCartItemMock = overrides => ({
  request: { query: CURRENT_USER_QUERY },
  result: {
    data: {
      me: {
        ...mockUser,
        cart: [{
          ...mockCartItem,
          ...overrides
        }],
      },
    },
  },
});

const signoutMutationMock = {
  request: { query: SIGNOUT_MUTATION },
  result: {
    data: {
      signout: {
        __typename: 'Message',
        success: true,
        message: 'Goodbye!'
      },
    },
  },
};

const requestPasswordResetMutationMock = {
  request: { query: REQUEST_PASSWORD_RESET_MUTATION, variables: { email: 'answart@sbcglobal.net' } },
  result: {
    data: {
      requestPasswordReset: {
        __typename: 'Message',
        success: true,
        message: 'Your request has been sent! Check your email.',
      }
    },
  },
};

const requestPasswordResetMutationErrorMock = {
  request: { query: REQUEST_PASSWORD_RESET_MUTATION, variables: { email: 'err@g.c' } },
  result: {
    data: {
      requestPasswordReset: {
        __typename: 'Message',
        success: false,
        message: 'No such user found for email err@g.c',
      }
    },
  },
};


export {
  mockUser,
  userQueryEmptyCartMock, userQueryNoUserMock, userQueryCartItemMock,
  signoutMutationMock,
  requestPasswordResetMutationMock, requestPasswordResetMutationErrorMock
};