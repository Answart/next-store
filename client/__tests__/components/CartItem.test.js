import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import CartItem from '../../components/CartItem';
import { mockCartItem, mockVariant } from '../../lib/test-utils/mocks';


describe('<CartItem />', () => {
  afterAll(() => wrapper.unmount());

  it('renders with proper data when loading w/out variant', async () => {
    const wrapper = mount(
      <MockedProvider mocks={[]}>
        <CartItem id={mockCartItem.id} quantity={mockCartItem.quantity} variant={null} />
      </MockedProvider>
    );
    wrapper.update();
    expect(wrapper.find('.cart-item-details').text()).toContain('This Product has been removed');
    expect(wrapper.find('RemoveFromCart').props().id).toContain(mockCartItem.id);
    wrapper.unmount();
  });

  describe('when loading w/variant', () => {
    let wrapper;
    beforeAll(() => {
      wrapper = mount(
        <MockedProvider mocks={[]}>
          <CartItem id={mockCartItem.id} quantity={mockCartItem.quantity} variant={mockVariant} />
        </MockedProvider>
      );
    });
    afterAll(() => {
      wrapper.unmount();
      jest.clearAllMocks();
    });

    it('renders all extra components properly', async () => {
      expect(wrapper.find('Link').length).toBe(1);
      expect(wrapper.find('PriceTag').length).toBe(1);
      expect(wrapper.find('UpdateCartItem').length).toBe(2);
      expect(wrapper.find('RemoveFromCart').length).toBe(1);
      expect(toJSON(wrapper.find('Header'))).toMatchSnapshot();
    });

    it('renders properly', async () => {
      expect(wrapper.find('Link').props().href).toMatchObject({
        pathname: "/shop",
        query: { id: mockVariant.product.id }
      });
      const details = wrapper.find('.cart-item-detail');
      expect(details.length).toBe(4);
      expect(details.at(0).text()).toBe(mockVariant.product.title);
      expect(details.at(1).text()).toBe(`$${mockVariant.price}$${mockVariant.salePrice}`);
      expect(details.at(2).text()).toBe(`Size: ${mockVariant.size}`);
      expect(details.at(3).text()).toBe('Color: White');
      const actions = wrapper.find('.cart-item-quantity-actions');
      expect(actions.text()).toBe('3Remove');
      expect(actions.text()).toBe(`${mockCartItem.quantity}Remove`);
      const price = wrapper.find('.cart-item-total-price');
      expect(price.text()).toBe(`$${mockCartItem.quantity * mockVariant.salePrice}`);
    });
  });
});
