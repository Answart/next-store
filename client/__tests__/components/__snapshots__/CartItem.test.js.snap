// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartItem /> when loading w/variant renders all extra components properly 1`] = `null`;
